test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,231 @@
+
+	// Sanity check: Test bank send before upgrade
+	s.T().Log("Testing bank send functionality before upgrade")
+	testBankSend(s.T(), chain, cfg)
+
+	validatorNode := chain.GetNodes()[0]
+	rpcClient, err := validatorNode.GetRPCClient()
+	s.Require().NoError(err, "failed to get RPC client")
+
+	abciInfo, err := rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+	currentAppVer := abciInfo.Response.GetAppVersion()
+	s.T().Logf("Current app version before upgrade: %d", currentAppVer)
+
+	records, err := kr.List()
+	s.Require().NoError(err, "failed to list accounts")
+	s.Require().Len(records, len(chain.GetNodes()), "number of accounts does not match number of nodes")
+
+	// Signal for the upgrade to the new version
+	for i, node := range chain.GetNodes() {
+		s.T().Logf("Signaling for upgrade to version %d from validator %d", upgradeCfg.TargetAppVersion, i)
+
+		signalCmd := []string{"tx", "signal", "signal", fmt.Sprintf("%d", upgradeCfg.TargetAppVersion), "--from", records[i].Name}
+		_, stderr, err := s.ExecuteNodeCommand(ctx, node, signalCmd...)
+		s.Require().NoError(err, "failed to signal for upgrade: %s", stderr)
+	}
+
+	s.Require().NoError(wait.ForBlocks(ctx, 2, chain))
+
+	s.validateSignalTally(ctx, validatorNode, upgradeCfg.TargetAppVersion)
+
+	// Execute try-upgrade transaction on all nodes
+	for i, node := range chain.GetNodes() {
+		s.T().Logf("Executing try-upgrade transaction on validator %d", i)
+		tryUpgradeCmd := []string{"tx", "signal", "try-upgrade", "--from", records[i].Name}
+		_, upgradeStderr, err := s.ExecuteNodeCommand(ctx, node, tryUpgradeCmd...)
+		s.Require().NoError(err, "failed to execute try-upgrade on validator %d: %s", i, upgradeStderr)
+	}
+
+	s.T().Log("Querying upgrade info")
+	upgradeInfoCmd := []string{"query", "signal", "upgrade", "--output", "json"}
+	upgradeInfoStdout, upgradeInfoStderr, err := s.ExecuteNodeCommand(ctx, validatorNode, upgradeInfoCmd...)
+	s.Require().NoError(err, "failed to query upgrade info: %s", upgradeInfoStderr)
+	s.T().Logf("Upgrade info: %s", upgradeInfoStdout)
+
+	// Wait for the upgrade to be scheduled
+	s.Require().NoError(wait.ForBlocks(ctx, 5, chain))
+
+	err = chain.Stop(ctx)
+	s.Require().NoError(err)
+
+	chain.UpgradeVersion(ctx, upgradeCfg.TargetVersion)
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
+
+	err = wait.ForBlocks(ctx, 2, chain)
+	s.Require().NoError(err)
+
+	// Verify the version after upgrade
+	rpcClient, err = validatorNode.GetRPCClient()
+	s.Require().NoError(err, "failed to get RPC client for version check")
+
+	abciInfo, err = rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+
+	// The version string might vary, but should contain the commit hash
+	var (
+		versionStr     = abciInfo.Response.GetVersion()
+		trimmedVersion = strings.TrimPrefix(upgradeCfg.TargetVersion, "v")
+	)
+	s.Assert().Contains(versionStr, trimmedVersion, "version should contain %q", trimmedVersion)
```

Comment:
there's still a mix of s.Assert and s.Require and makes things read weirdly, can you make them consistent (my preference is for Require)

Reply:
I see this line failing with `"5378ef3" does not contain "5378ef3e"`, maybe that was an older run of the test?

Reply:
Ok I'll fix the required thing. I was assuming required was usually used for the code that prepares the evaluation and assert is for the actual evaluation part. but ok I will make them all to use required.

About the hash mismatch, there was an issue in the gh workflow, now it is fixed.

Reply:
awesome thanks! Ah yeah I guess I can see the mix of Require/Assert making sense in that context. It is not a big deal either way I think.